### PR TITLE
Support for multiple upstreams

### PR DESCRIPTION
**Pull request by prilka**
I have added support for multiple upstreams so that Loadbalancing or HA is possible.

A few changes to the templates were necessary. I tried to make the changes as backward-compatible
as possible for people who use custom templates.

### CODE REVIEW

**Comment by SteveLTN** [comment]
Good job! I absolutely appreciate the your initiative and the effort!

However, I'm afraid I'm not going to merge this PR. Here is why.

When I created this project, I wanted a quick and easy way of setting up an HTTPS powered website. I knew Let's Encrypt but found the toolchain hard to use, especially for beginners, as it depends on your existing setup. Therefore I created this project and used a simple syntax, configured through ENV variables.

The project grew since then, and every time when someone asks me to add a feature, I would evaluate if it's worth the effort. And to me, load-balancing is a complex topic, you'll configure your system differently based on your own scenario. I don't feel like shoveling all configuration bits into one single ENV variable is a good idea. 

When users have requirements such as load-balancing, clearly they are going beyond the "simple website" scenario. In this case, I'm sure the size of the project will grow, the configuration will become more complex. And shoveling complex configuration into a single-line ENV variable isn't a good idea. Maybe with two or three upstreams the ENV is still readable, but when it grows, you'll eventually need a cleaner, more manageable way of organizing your config. I think such users are better off just take out the Nginx config and do however complex configurations they want through it. I feel like expanding the ENV config syntax is just re-inventing Nginx config language, in a terrible way.

Hope that explains my thoughts. Again I appreciate your effort and feel sorry to disappoint you for not being able to merge your PR.

**Comment by prilka** [comment]
Thanks for your quick reply. But it's a little disappointing. It is only a small code change and a small extension of the syntax.

I have a slightly different view on HTTPS-Portal. I see the project as an easy replacement for tools like traefik for small to medium websites, but with a much cleaner and simpler configuration. 

I work in an agency where we manage hundreds of websites. Since we started using HTTPS-Portal, the configuration and maintenance effort has decreased dramatically. Thanks for that. In combination with the multiline string features of yaml (e.g. `DOMAINS: >`) a very simple and clear configuration is possible even with a high number of domains.
For some installations we habe 20+ (sub-)domains configured. The configuration and maintainability is much easier than with traefik or vanilla nginx.

The use case I currently have has nothing to do with high-traffic sites. A small/medium sized website should get an additional sub-domain. The upstream is an external cluster with two IPs. HTTPS-Portal does not offer me an easy way to configure this use case and I don't want to write an nginx config and maintain it for such a small use case.

Therefore I think that this PR adds another useful feature to HTTPS-Portal. But in the end you are the maintainer of the project and I respect that. If you don't want to merge this PR I have to deal with it.

**Comment by SteveLTN** [comment]
All right. You have convinced me.
I'll have a more detailed look maybe in the weekend, and do some testing myself.

File: fs_overlay/opt/certs_manager/models/domain.rb
Comment (SteveLTN) [inline fs_overlay/opt/certs_manager/models/domain.rb:103]: I guess this is for backwards compatibility?

File: fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb
Comment (SteveLTN) [inline fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb:4]: Let's call it `upstream[:address]`, shall we? 
We have too many occurrences of the word "upstream" here.
And the [Nginx doc](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#upstream) calls it `address`: 
> Syntax:	server address [parameters];

**Review by SteveLTN — COMMENTED** [review]
Looks pretty good. If you would fix the comments, I'll merge it then do a release after some tests (probably in the weekend though).

**Comment by prilka** [comment]
> All right. You have convinced me.
> I'll have a more detailed look maybe in the weekend, and do some testing myself.

Thanks for changing your mind. I will check your remarks and update the PR.

File: fs_overlay/opt/certs_manager/models/domain.rb
Comment (prilka) [reply fs_overlay/opt/certs_manager/models/domain.rb:103]: Yes. It's important to return nil for static-pages and redirect-mode otherwise the condition in default.ssl.conf.erb would be broken. I will add a comment.

File: fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb
Comment (prilka) [reply fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb:4]: Good point. `address` is a much better name.

File: fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb
Comment (prilka) [reply fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb:1]: I wasn't sure if the upstream block behaves exactly the same for the `DYNAMIC_UPSTREAM` case. Are you sure that both behave the same?

However, some kind of condition is necessary in any case. For the static site and redirect-mode the upstream block makes no sense.
An empty upstream block causes the following nginx error: `nginx: [emerg] no servers are inside upstream ...`

I could check for `domain.upstream` here and simplify the proxy_pass part later.

Comment (SteveLTN) [reply fs_overlay/var/lib/nginx-conf/default.ssl.conf.erb:1]: I wasn't sure about `DYNAMIC_UPSTREAM`. Okay, leave it as it is. It's not that complex anyway.

**Review by SteveLTN — APPROVED** [review]

**Comment by SteveLTN** [comment]
1.14.0 is released!

**Comment by prilka** [comment]
Nice, thank you!

**Pull request merged**
